Fix node tool cleanup

henningandersen opened the pull request:
In case of failures, node tool could end up removing up both old
and new cluster state.

henningandersen committed:
Fix node tool cleanup

In case of failures, node tool could end up removing up both old
and new cluster state.

elasticmachine commented:
Pinging @elastic/es-distributed

henningandersen commented:
@elasticmachine run elasticsearch-ci/2

andrershov reviewed (changes requested):
@henningandersen It would be nice to add a test case for this change - old metadata cleanup fails, but the tool execution is successful and we can read new metadata. Looks good otherwise.

henningandersen committed:
Fix node tool cleanup

Added test case that checks that we leave new meta data behind if old
meta data cleanup fails.

henningandersen committed:
Merge remote-tracking branch 'origin/master' into fix_node_tool_metadata_cleanup

henningandersen commented:
>  It would be nice to add a test case for this change - old metadata cleanup fails, but the tool execution is successful and we can read new metadata.

Added in ed3812ce7b64c1021439fc282114e1f530cbc701

Thanks @andrershov

andrershov reviewed (changes requested):
@henningandersen thank you for adding a test for this. I think that instead of adding a custom method that looks for all metadata files, we can use `MetaDataStateFormat.loadLatestStateWithGeneration` and `MetaDataStateFormat.loadGeneration` for `Manifest` and `Metadata` to check that the first run of the tool will not delete new generation. And the next run of the tool will delete both previous generations and create a new one.

henningandersen committed:
Fix node tool cleanup

Improved test case.

henningandersen commented:
> we can use `MetaDataStateFormat.loadLatestStateWithGeneration` and `MetaDataStateFormat.loadGeneration` for `Manifest` and `Metadata` to check that the first run of the tool will not delete new generation.

Thanks @andrershov , fixed in 9ff3b35df4442c4507ebfdb7013487535790d628, please have another look.

andrershov commented on server/src/test/java/org/elasticsearch/cluster/coordination/ElasticsearchNodeCommandIT.java:
```diff
@@ -418,4 +423,57 @@ public void testCanRunUnsafeBootstrapAfterErroneousDetachWithoutLoosingMetaData(
+        executeCommand(new UnsafeBootstrapMasterCommand(), environment, 0, false);
+
+        assertNull(loadMetaData(dataPaths, namedXContentRegistry, originalManifest));
+        assertNull(loadMetaData(dataPaths, namedXContentRegistry, secondManifest));
```
Can we also check that loadLatestManifest and loadMetadata will load non-null objects?

henningandersen replied:
Fixed in ca59377ea745834b2d1477a1cde5d62cfc68fe77

andrershov reviewed (approved):
LGTM. I left one small comment, after it's addressed, no need for the next round.

henningandersen committed:
Fix node tool cleanup

Improved test case.
